# Processing of dual RNA-seq data for the PhageExpressionAtlas

This repo contains the code for processing dual RNA-seq data as currently hosted in the PhageExpressionAtlas (**LINK**). First, a nextflow pipeline is used to trim paired-end or single-end RNA-seq reads, check their quality, map the reads to a dual genome, filter the alignments and count reads. A Jupyter Notebook is then used to annotate pre-process the count data for storage and analysis with the PhageExpressionAtlas.

## Nextflow pipeline


### Tools in the pipeline


### Installation & usage


### Input & parameters


### Output


## Pre-processing with the Juypter notebook


### Steps


### Output


## Customization of workflow



## Contributing to the PhageExpressionAtlas


## References & Acknowledgements